<div align="center">

# Справочник по NumPy

</div>

---

**NumPy** — это библиотека для работы с данными и машинного обучения. Она позволяет создавать массивы и быстро их обрабатывать, так как выполняет операции не поэлементно, а сразу над всем массивом. Благодаря реализации на чистом **C**, **NumPy** обеспечивает низкоуровневую оптимизацию и значительно выигрывает по скорости по сравнению с обычными массивами **Python**.

In [3]:
import numpy as np

<div align="center">

# 1. Одномерные массивы

</div>

---


In [4]:
a = [1, 2, 3] # Классический python список
b = np.array(a, dtype = 'float64') # NumPy список
print(type(a), type(b))

<class 'list'> <class 'numpy.ndarray'>


In [5]:
b # явно указали что данные должны быть вещественного типа

array([1., 2., 3.])

In [6]:
a = [1, 2, 'c'] # Python позволяет в списках хранить разные типы,
# для NumPy важно, чтобы все элементы массивы были однородны

b = np.array(a)
print('Для лист:', type(a[0]),
      '\nДля np.array:', type(b[0]))

Для лист: <class 'int'> 
Для np.array: <class 'numpy.str_'>


---

В массивах **NumPy** все элементы должны быть одного типа. В отличие от списков **Python**, где это не имеет значения, в **NumPy** желательно соблюдать однородность: если в массиве встречаются разные типы данных, произойдёт неявное приведение (каст) всех элементов к одному типу.

Например, при создании массива `np.array` из чисел и строк, все элементы будут преобразованы в строки. Это связано с иерархией типов данных в **NumPy**: строка имеет более высокий приоритет, чем целое число, так как число всегда можно привести к строке, а строку — не всегда к числу.

---

In [7]:
# Также в конструктор можно сразу передавать сразу список 
b = np.array([1, 2, 3])
print(type(b))

<class 'numpy.ndarray'>


### методы класса ndarray

Возьмем все методы объекта b с помошью команды `dir` и вычтем все объекты класса python `object`, чтобы увидеть какие есть методы класса **ndarray**.

set(dir(b)) - set(dir(object))

In [8]:
# Рассмотрим метод nbytes
arr = np.array([1, 2, 3, 4, 5], dtype = 'int32') 

# int32 - это 32 бита, то есть 4 байта
# 4 байта * 5 элементов нашего массива = 20
arr.nbytes

20

### Рассмотрим количественные характеристики ndarray

In [9]:
# Создадим 3-х мерный массив для примера
arr = np.array([[[1, 2, 3, 6],
                 [4, 5, 6, 8],
                 [7, 8, 9, 1]],
                 [[1, 2, 3, 5],
                 [4, 5, 6, 6],
                 [7, 8, 9, 4]]])
print(arr)

[[[1 2 3 6]
  [4 5 6 8]
  [7 8 9 1]]

 [[1 2 3 5]
  [4 5 6 6]
  [7 8 9 4]]]


In [10]:
# Количественные хар-ки
print("Длина:", len(arr), "-- кол-во элементов по первой оси.",
      "\nsize:", arr.size, "-- всего элементов в матрице.",
      "\nndim:", arr.ndim, "-- рамзерность матрицы",
      "\nshape:", arr.shape, "-- кол-во элементов по каждой оси")



Длина: 2 -- кол-во элементов по первой оси. 
size: 24 -- всего элементов в матрице. 
ndim: 3 -- рамзерность матрицы 
shape: (2, 3, 4) -- кол-во элементов по каждой оси


---

* **`len()`** — встроенная функция Python. Для массива NumPy она возвращает количество элементов вдоль **первой оси**. Например, трёхмерный массив можно рассматривать как массив двумерных массивов, и `len()` покажет, сколько таких двумерных блоков содержится.
* **`size`** — общее количество элементов во всём массиве.
* **`ndim`** — количество измерений (осей) массива. В данном случае массив трёхмерный.
* **`shape`** — кортеж, описывающий количество элементов вдоль каждой оси. Например, `(2, 3, 4)` означает, что массив состоит из **2** матриц размером `3 × 4`.

---

### Индексы

In [11]:
a = np.array([1, 2, 4, 5]) 
a[0], a[1] # индексация работает также как и у Python

(np.int64(1), np.int64(2))

In [12]:
a[-1] # Выводит последний элемент

np.int64(5)

In [13]:
a[-2] # Выводит предпоследний элемент и тд

np.int64(4)

In [14]:
# Можем изменить элемент массива по индексу 
a[1] = -2
a

array([ 1, -2,  4,  5])

---

`ndarray` можно использовать в циклах, однако это нежелательно, поскольку в таком случае теряется главное преимущество **NumPy** — высокое быстродействие. Вместо перебора элементов лучше применять векторизованные операции, которые работают сразу со всем массивом.

---

In [15]:
# Цикл по массиву
for i in a:
    print(i)

1
-2
4
5


### Создание массивов

In [16]:
a = np.zeros(7) # Массив из нулей, по умолчнанию будут float64
b = np.ones(7, dtype = np.int16) # Массив из едениц
print(a)
print(b)

[0. 0. 0. 0. 0. 0. 0.]
[1 1 1 1 1 1 1]


#### Создание нулевых или еденичных массивов размером как другой

In [17]:
arr = np.array([1, 2, 4, 6], dtype = 'int32')
zero = np.zeros_like(arr)
print(zero)

[0 0 0 0]


In [18]:
arr = np.array([1, 2, 4, 6], dtype = 'int32')
zero = np.ones_like(arr)
print(zero)

[1 1 1 1]


> Метод `np.zeros_like` и `np.ones_like` берет не только форму ихсодного массива, а также его тип данных.

---

Функция `np.arange` работает аналогично встроенной функции `range` в Python и позволяет создавать последовательность чисел с фиксированным шагом. Она принимает до трёх аргументов:

1. **start** — число, с которого начинается последовательность (по умолчанию `0`);
2. **stop** — число, до которого идёт последовательность (не включая его);
3. **step** — шаг, с которым формируется последовательность (по умолчанию `1`).

Таким образом:

* при одном аргументе указывается только **stop**, а начало считается равным `0`, шаг — `1`;
* при двух аргументах задаются **start** и **stop**, шаг остаётся равным `1`;
* при трёх аргументах задаются **start**, **stop** и **step** явно.

---

In [19]:
a = np.arange(1, 16, 4)
b = np.arange(5., 21, 2)
c = np.arange(1, 10)
d = np.arange(5)
print(a)
print(b)
print(c)
print(d)

[ 1  5  9 13]
[ 5.  7.  9. 11. 13. 15. 17. 19.]
[1 2 3 4 5 6 7 8 9]
[0 1 2 3 4]


---

Функция `np.linspace` используется для создания равномерно распределённых чисел на заданном отрезке. В отличие от `np.arange`, здесь указывается не шаг, а количество элементов.

* **первый аргумент** — начало диапазона (включается);
* **второй аргумент** — конец диапазона (включается);
* **третий аргумент** — количество точек, которые нужно сгенерировать.

Таким образом, `np.linspace(0, 1, 5)` создаст массив `[0. , 0.25, 0.5 , 0.75, 1. ]`.

---

In [20]:
a = np.linspace(1, 15, 2)
b = np.linspace(5, 12, 10)
print(a)
print(b)

[ 1. 15.]
[ 5.          5.77777778  6.55555556  7.33333333  8.11111111  8.88888889
  9.66666667 10.44444444 11.22222222 12.        ]


In [21]:
# Пример: вывести числа от 10 до 32 12 чисел и узнать шаг.
a = np.linspace(10, 32, 12)
print(a)
print('Шаг:', a[1] - a[0])

[10. 12. 14. 16. 18. 20. 22. 24. 26. 28. 30. 32.]
Шаг: 2.0


---

Функция `np.logspace` создаёт последовательность чисел с постоянным **логарифмическим шагом**.
Аргументы:

1. **start** — нижняя граница (указывается как показатель степени);
2. **stop** — верхняя граница (также как показатель степени);
3. **num** — количество точек в последовательности.

По умолчанию основание степени равно `10`, но его можно изменить с помощью параметра `base`.

---

In [22]:
a = np.logspace(0, 3, 12)
print(a)

[   1.            1.87381742    3.51119173    6.57933225   12.32846739
   23.101297     43.28761281   81.11308308  151.9911083   284.80358684
  533.66992312 1000.        ]


<div align="center">

# 2. Операции над одномерными массивами

</div>

---

Все арифметические операции производятся поэлементно

In [23]:
a = np.linspace(3, 33, 11)
b = np.linspace(-2, -22, 11)

print(a)
print(b)
print('\n\n')
print(a + b)
print(a - b)
print(a * b)
print(a / b)

[ 3.  6.  9. 12. 15. 18. 21. 24. 27. 30. 33.]
[ -2.  -4.  -6.  -8. -10. -12. -14. -16. -18. -20. -22.]



[ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11.]
[ 5. 10. 15. 20. 25. 30. 35. 40. 45. 50. 55.]
[  -6.  -24.  -54.  -96. -150. -216. -294. -384. -486. -600. -726.]
[-1.5 -1.5 -1.5 -1.5 -1.5 -1.5 -1.5 -1.5 -1.5 -1.5 -1.5]


**Один из операндов может быть скаляром, а не массивом**

In [24]:
print(5 * a)
print(10 + b)

[ 15.  30.  45.  60.  75.  90. 105. 120. 135. 150. 165.]
[  8.   6.   4.   2.   0.  -2.  -4.  -6.  -8. -10. -12.]


In [25]:
print((a + b) ** 2)
print(2 ** (a + b))

[  1.   4.   9.  16.  25.  36.  49.  64.  81. 100. 121.]
[2.000e+00 4.000e+00 8.000e+00 1.600e+01 3.200e+01 6.400e+01 1.280e+02
 2.560e+02 5.120e+02 1.024e+03 2.048e+03]


2.048e+03 расшифровывается как 2.048 * $10^3$

**Если типы элементов разные, то идет каст к большему**

In [26]:
a = np.linspace(3, 33, 11) # float64
b = a + np.arange(11, dtype = 'int16')
print(type(b[0]))

<class 'numpy.float64'>


---

Помимо арифметических операций, в **NumPy** существуют элементарные функции, которые называются **универсальными функциями** (*ufunc*). К ним относятся: синус, косинус, экспонента, логарифм, а также обратные тригонометрические функции и многие другие.

Как и арифметические операции, универсальные функции применяются **поэлементно** ко всем элементам массива.

---

In [27]:
type(np.cos)

numpy.ufunc

In [28]:
a = np.linspace(3, 33, 11)
b = np.linspace(-2, -22, 11)
print(a)
print(b)

[ 3.  6.  9. 12. 15. 18. 21. 24. 27. 30. 33.]
[ -2.  -4.  -6.  -8. -10. -12. -14. -16. -18. -20. -22.]


In [29]:
np.cos(a)

array([-0.9899925 ,  0.96017029, -0.91113026,  0.84385396, -0.75968791,
        0.66031671, -0.54772926,  0.42417901, -0.29213881,  0.15425145,
       -0.01327675])

In [30]:
np.log(b)

/tmp/ipykernel_3585/4200754868.py:1: RuntimeWarning: invalid value encountered in log
  np.log(b)


array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan])

---

Мы не можем взять логарифм от массива `b`, так как он содержит отрицательные числа. Логарифм для отрицательных значений в действительной области **не определён**, поэтому результатом становятся значения `nan` (*not a number*).

---

**Логические операции также производятся поэлементно**

In [31]:
print(a > b)
print(a == b)
print(a < b)
print(a >= 10)
print(a <= 10)

[ True  True  True  True  True  True  True  True  True  True  True]
[False False False False False False False False False False False]
[False False False False False False False False False False False]
[False False False  True  True  True  True  True  True  True  True]
[ True  True  True False False False False False False False False]


---

**Квантор существования** (∃) — утверждает, что **существует хотя бы один элемент** множества, для которого выполняется данное условие или логическое выражение.
Пример:

$$
\exists x \in A : x > 0
$$

— в множестве $A$ есть хотя бы один положительный элемент.

**Квантор всеобщности** (∀) — утверждает, что **для всех элементов множества** выполняется некоторое условие.
Пример:

$$
\forall x \in A : x \geq 0
$$

— все элементы множества $A$ неотрицательные.

---

In [32]:
c = np.arange(0., 20)
print(c)
print(type(c[0]))

[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.
 18. 19.]
<class 'numpy.float64'>


In [33]:
# Для всеобщности
print(np.all(c != 0.))

False


In [34]:
# Для существования
print(np.any(c == 0.))

True


---

**Inplace-операции** — это операции, которые изменяют существующий объект, не создавая новый. В NumPy такие операции возможны только тогда, когда операнды совместимы по типу данных (dtype).

---

In [35]:
c = np.arange(0., 20)
c += np.sin(4)
print(c)

[-0.7568025  0.2431975  1.2431975  2.2431975  3.2431975  4.2431975
  5.2431975  6.2431975  7.2431975  8.2431975  9.2431975 10.2431975
 11.2431975 12.2431975 13.2431975 14.2431975 15.2431975 16.2431975
 17.2431975 18.2431975]


In [36]:
c *= 2
print(c)

[-1.51360499  0.48639501  2.48639501  4.48639501  6.48639501  8.48639501
 10.48639501 12.48639501 14.48639501 16.48639501 18.48639501 20.48639501
 22.48639501 24.48639501 26.48639501 28.48639501 30.48639501 32.48639501
 34.48639501 36.48639501]


**При делении ndarray на нули, исключение не бросается**

In [37]:
a = np.array([0.0, 0.0, 1.0, -1.0])
b = np.array([1.0, 0.0, 0.0, 0.0])
print(a / b, 
      'Получится просто 0, неопределенное число, плюс бесконечность, '
      'и минус бесконечность')

[  0.  nan  inf -inf] Получится просто 0, неопределенное число, плюс бесконечность, и минус бесконечность


/tmp/ipykernel_3585/3090575395.py:3: RuntimeWarning: divide by zero encountered in divide
  print(a / b,
/tmp/ipykernel_3585/3090575395.py:3: RuntimeWarning: invalid value encountered in divide
  print(a / b,


**Математические константы**

In [38]:
print(np.e)
print(np.pi)

2.718281828459045
3.141592653589793


---

**Кумулятивная сумма** — это последовательность, в которой каждый элемент равен сумме всех предыдущих элементов исходного массива, включая текущий.

---

In [39]:
b = np.arange(1., 21, 1)
c = b.cumsum()
print(b)
print(c)

[ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17. 18.
 19. 20.]
[  1.   3.   6.  10.  15.  21.  28.  36.  45.  55.  66.  78.  91. 105.
 120. 136. 153. 171. 190. 210.]


**Сортировка в NumPy массивах**

In [40]:
# Первый метод сортирует, но не изменяет его
a = np.array([9, -4, 0, 1, 5])
print(np.sort(a))
print(a)

[-4  0  1  5  9]
[ 9 -4  0  1  5]


In [41]:
# Второй метод сортирует и изменяет его
a = np.array([9, -4, 0, 1, 5])
a.sort()
print(a)

[-4  0  1  5  9]


**Объединение массивов**

In [42]:
a = np.array([9, -4, 0, 1, 5])
b = np.ones(5)
c = np.hstack((a, b, b*5))
c

array([ 9., -4.,  0.,  1.,  5.,  1.,  1.,  1.,  1.,  1.,  5.,  5.,  5.,
        5.,  5.])

**Расщепление массивов**

In [43]:
a = np.array([9, -4, 0, 1, 5, 9, 10])
x1, x2, x3, x4 = np.hsplit(a, [3, 5, 6]) # Здесь указываем разделять до индекса
print(a)
print(x1) # Разделит до 3 индекса
print(x2) # Разделит от 3 до 5 индекса
print(x3) # Разделит от 5 до 6 индекса
print(x4) # то что останется

[ 9 -4  0  1  5  9 10]
[ 9 -4  0]
[1 5]
[9]
[10]


**У NumPy функции `append`, `insert`, `delete` не являются inplace: они создают новый массив и возвращают его.**

* `np.append(arr, values)` — добавляет элементы в конец массива.
* `np.insert(arr, index, values)` — вставляет элементы по указанному индексу.
* `np.delete(arr, index)` — удаляет элементы по индексу.

In [46]:
a = np.array([9, -4, 0, 1, 5, 9, 10])
print(np.delete(a, [3, 4, 1]))
print(a)

[ 9  0  9 10]
[ 9 -4  0  1  5  9 10]


In [47]:
a = np.array([9, -4, 0, 1, 5, 9, 10])
print(np.insert(a, 2, [-1, -1]))

[ 9 -4 -1 -1  0  1  5  9 10]


In [48]:
a = np.array([9, -4, 0, 1, 5, 9, 10])
print(np.append(a, [0, 0, 0]))

[ 9 -4  0  1  5  9 10  0  0  0]


---

**Индексирование массива и срезы в NumPy**

1. Массив в обратном порядке;
2. Диапазоны индексов: левая граница включается, правая — нет; при этом создаётся представление, а не новый массив;
3. Диапазоны индексов с шагом (по умолчанию 1): левая граница включается, правая — нет;
4. Важный момент: срез не создаёт копию, а является представлением исходного массива и указывает на ту же область памяти;
5. Подмассиву можно присвоить скаляр.

---

In [53]:
# 1)
a = np.array([9, -4, 0, 1, 5, 9, 10])
a[::-1]

array([10,  9,  5,  1,  0, -4,  9])

In [54]:
# 2)
a = np.array([9, -4, 0, 1, 5, 9, 10])
a[2:5]

array([0, 1, 5])

In [55]:
# 3) 
a = np.array([9, -4, 0, 1, 5, 9, 10])
a[1:6:2]

array([-4,  1,  9])

In [58]:
# 4)
a = np.array([9, -4, 0, 1, 5, 9, 10])
b = a[0:6]
b[1] = -1000
print(a)

[    9 -1000     0     1     5     9    10]


In [61]:
# 5)
a = np.array([9, -4, 1, 1, 5, 9, 10])
a[1:6:3] = 0
print(a)

[ 9  0  1  1  0  9 10]


---

**Копирование массива в отдельную область памяти**
С помощью метода `copy` можно создать независимую копию массива.
В этом случае новый массив будет храниться в другой области памяти и не будет связан с исходным.

Кроме того, в отличие от индексов в Python, в NumPy можно передавать массив индексов.

---

In [64]:
a = np.array([9, -4, 1, 1, 5, 9, 10])
b = a.copy()
b[2] = -9
print(b)
print(a)

[ 9 -4 -9  1  5  9 10]
[ 9 -4  1  1  5  9 10]


In [67]:
# Несколько индексов
a = np.array([9, -4, 1, 1, 5, 9, 10])
print(a[[1, 3, 5]])

[-4  1  9]


<div align="center">

# 3. Двумерные массивы

</div>

---

**Двумерные массивы создаются так же, как и одномерные, но на вход передаётся список списков — по сути, матрица.**

In [83]:
a = np.array([[1, 2], [3, 4]])
print(a)

[[1 2]
 [3 4]]


In [84]:
# Кол-во осей, размерность = 2, так как двумерный массив у нас
a.ndim

2

In [85]:
# Разерность 2 строки и  2 столбца
a.shape

(2, 2)

In [86]:
# Кол-во элементов
a.size

4

In [87]:
# Длина, число элементов по первой оси, тоже = 2
len(a)

2

---

**Индексация в двумерном массиве**

Существует два способа:

1. Аналогично спискам в Python — сначала указываем индекс строки, затем столбца, через двойные квадратные скобки (`a[0][1]`);
2. Использовать одну пару квадратных скобок и перечислять индексы через запятую (`a[0, 1]`).

---

In [88]:
# 1)
a = np.array([[1, 2], [3, 4]])
a[1][1]

np.int64(4)

In [89]:
# 2)
a = np.array([[1, 2], [3, 4]])
a[1, 1]

np.int64(4)

---

**`shape`**

Атрибут `shape` позволяет изменять размерность массива без изменения самих данных.
Размерность задаётся в виде кортежа. При этом общее число элементов в новом представлении должно совпадать с числом элементов исходного массива.

---

In [92]:
a = np.arange(0, 20)
a.shape = (2, 10)
print(a)

[[ 0  1  2  3  4  5  6  7  8  9]
 [10 11 12 13 14 15 16 17 18 19]]


---

**Метод `ravel`**

Метод `ravel` преобразует многомерный массив в одномерный, «разворачивая» все его элементы в последовательность.

---

In [93]:
a = np.array([[1, 2], [3, 4]])
print(a.ravel())

[1 2 3 4]


---

Так же, как в одномерном массиве, в двумерном можно создавать матрицы из нулей или единиц, передавая в качестве аргумента кортеж размерности `(n — число строк, m — число столбцов)`.

---

In [95]:
a = np.zeros((3, 3))
print(a)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]


In [96]:
a = np.ones((3, 4))
print(a)

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]


---

Функция `eye()` — важный инструмент для линейной алгебры, так как она создаёт квадратную единичную матрицу.

Также можно создавать произвольные матрицы с ненулевыми элементами только на главной диагонали с помощью функции `diag()`. На вход передаётся массив элементов, которые будут расположены на диагонали матрицы.

---

In [99]:
# Еденичная матрица
a = np.eye(3)
print(a)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [100]:
# Произвольная матрица по диагонали
a = np.diag(np.array([1, 2, 3, 4]))
print(a)

[[1 0 0 0]
 [0 2 0 0]
 [0 0 3 0]
 [0 0 0 4]]


In [104]:
# Пример
# Создать квадратную матрицу размером (8), где на главной 
# диагонали будет арифметическая прогрессия с шагом 3, начиная с 3,
# а на побочной диагонали -1.

a = -1 * np.eye(8)[::-1] + np.diag(np.arange(3, 27, 3))
print(a)

[[ 3.  0.  0.  0.  0.  0.  0. -1.]
 [ 0.  6.  0.  0.  0.  0. -1.  0.]
 [ 0.  0.  9.  0.  0. -1.  0.  0.]
 [ 0.  0.  0. 12. -1.  0.  0.  0.]
 [ 0.  0.  0. -1. 15.  0.  0.  0.]
 [ 0.  0. -1.  0.  0. 18.  0.  0.]
 [ 0. -1.  0.  0.  0.  0. 21.  0.]
 [-1.  0.  0.  0.  0.  0.  0. 24.]]


---

#### Умножение матриц в NumPy

1. **Поэлементное умножение** — выполняется для двух матриц одинаковой размерности (операция `*`).
2. **Матричное умножение** — выполняется с помощью оператора `@` или метода `dot()`. Для этого число столбцов первой матрицы должно совпадать с числом строк второй.
   Важно: матричное умножение **не коммутативно** (в общем случае `A @ B ≠ B @ A`).
   Метод `dot()` применяется как к векторно-скалярным операциям, так и к умножению матриц.

---

In [109]:
a = 5 * np.ones((5, 5))
b = np.eye(5) + 1
print(a,"\n")
print(b)

[[5. 5. 5. 5. 5.]
 [5. 5. 5. 5. 5.]
 [5. 5. 5. 5. 5.]
 [5. 5. 5. 5. 5.]
 [5. 5. 5. 5. 5.]] 

[[2. 1. 1. 1. 1.]
 [1. 2. 1. 1. 1.]
 [1. 1. 2. 1. 1.]
 [1. 1. 1. 2. 1.]
 [1. 1. 1. 1. 2.]]


In [110]:
# 1)
print(a * b)

[[10.  5.  5.  5.  5.]
 [ 5. 10.  5.  5.  5.]
 [ 5.  5. 10.  5.  5.]
 [ 5.  5.  5. 10.  5.]
 [ 5.  5.  5.  5. 10.]]


In [111]:
# 2)
print(a @ b)

[[30. 30. 30. 30. 30.]
 [30. 30. 30. 30. 30.]
 [30. 30. 30. 30. 30.]
 [30. 30. 30. 30. 30.]
 [30. 30. 30. 30. 30.]]


In [112]:
# 2)
print(a.dot(b))

[[30. 30. 30. 30. 30.]
 [30. 30. 30. 30. 30.]
 [30. 30. 30. 30. 30.]
 [30. 30. 30. 30. 30.]
 [30. 30. 30. 30. 30.]]


---

**Создание двумерных массивов с помощью `meshgrid()`**

Метод `meshgrid()` позволяет создать двумерные массивы, зависящие от одного индекса:

$$
X_{ij} = u_j, \quad Y_{ij} = v_i
$$

Распространённый пример применения — создание сетки для построения 3D-графиков.

---

In [113]:
u = np.linspace(1, 2, 2)
v = np.linspace(4, 8, 3)
print(u)
print(v)

[1. 2.]
[4. 6. 8.]


In [114]:
x, y = np.meshgrid(u, v)
print(x, '\n')
print(y)

[[1. 2.]
 [1. 2.]
 [1. 2.]] 

[[4. 4.]
 [6. 6.]
 [8. 8.]]


---

**Замена атрибута `shape`**

Более удобная запись для изменения размерности массива — метод `reshape()`.

* Если передать один аргумент, двумерный массив «вытягивается» в одномерный.
* Если передать кортеж, можно задать новую размерность массива (при этом общее число элементов должно совпадать с исходным).

---

In [117]:
x = np.array([[1, 2], [1, 2], [1, 2]])
print(x, '\n')
print(x.reshape(6))

[[1 2]
 [1 2]
 [1 2]] 

[1 2 1 2 1 2]


---

#### Маски

Маска — это массив булевых значений (`True` или `False`), имеющий ту же размерность, что и исходный массив.

---

In [153]:
a = np.arange(20)
print(a % 3 == 0) # Выводится массив [True, False] 
print(a[a % 3 == 0]) # Передача маски в качестве индекса 

[ True False False  True False False  True False False  True False False
  True False False  True False False  True False]
[ 0  3  6  9 12 15 18]


---

#### След (trace)

След матрицы — это сумма элементов главной диагонали матрицы.

---

In [123]:
a = np.arange(20)
b = np.diag(a[a >= 10])
print(b, '\n')
print(np.trace(b))

[[10  0  0  0  0  0  0  0  0  0]
 [ 0 11  0  0  0  0  0  0  0  0]
 [ 0  0 12  0  0  0  0  0  0  0]
 [ 0  0  0 13  0  0  0  0  0  0]
 [ 0  0  0  0 14  0  0  0  0  0]
 [ 0  0  0  0  0 15  0  0  0  0]
 [ 0  0  0  0  0  0 16  0  0  0]
 [ 0  0  0  0  0  0  0 17  0  0]
 [ 0  0  0  0  0  0  0  0 18  0]
 [ 0  0  0  0  0  0  0  0  0 19]] 

145


<div align="center">

# 4. Тензоры (Многомерные массивы)

</div>

---

In [126]:
# Создадим многомерный массив
x = np.arange(64).reshape(8, 2, 4) 
print(x)

[[[ 0  1  2  3]
  [ 4  5  6  7]]

 [[ 8  9 10 11]
  [12 13 14 15]]

 [[16 17 18 19]
  [20 21 22 23]]

 [[24 25 26 27]
  [28 29 30 31]]

 [[32 33 34 35]
  [36 37 38 39]]

 [[40 41 42 43]
  [44 45 46 47]]

 [[48 49 50 51]
  [52 53 54 55]]

 [[56 57 58 59]
  [60 61 62 63]]]


In [127]:
# Кол-во матриц 8, 
x.shape

(8, 2, 4)

In [128]:
# Размерность, массив у нас 3-х мерный
x.ndim

3

In [129]:
# Кол-во элементов
x.size

64

---

**Операции по осям в многомерных массивах**

Особенность работы с двумерными и многомерными массивами в NumPy в том, что многие операции (сумма, среднее, медиана и т.д.) можно применять не только ко всему массиву целиком, но и по отдельным осям.

Рассмотрим метод `sum()` с параметром `axis`:

* `axis=0` — суммирование вдоль **первой оси** (обычно по столбцам). Результат сохраняет количество элементов по оставшимся осям.
* `axis=1` — суммирование вдоль **второй оси** (обычно по строкам).
* `axis=2` — суммирование вдоль **третьей оси** (для трёхмерных массивов — по «глубине»).
* `axis=(i, j, …)` — можно передавать кортеж осей; суммирование производится одновременно по нескольким осям, и результат имеет размерность оставшихся осей (может стать одномерным массивом).

---

In [133]:
x = np.arange(64).reshape(8, 2, 4)
print(np.sum(x), '\n') # Сумма всех элементов массива
print(np.sum(x, axis = 0), '\n') # Суммирование вдоль первой оси
print(np.sum(x, axis = 1), '\n')
print(np.sum(x, axis = 2), '\n')
print(np.sum(x, axis = (1, 2)))

2016 

[[224 232 240 248]
 [256 264 272 280]] 

[[  4   6   8  10]
 [ 20  22  24  26]
 [ 36  38  40  42]
 [ 52  54  56  58]
 [ 68  70  72  74]
 [ 84  86  88  90]
 [100 102 104 106]
 [116 118 120 122]] 

[[  6  22]
 [ 38  54]
 [ 70  86]
 [102 118]
 [134 150]
 [166 182]
 [198 214]
 [230 246]] 

[ 28  92 156 220 284 348 412 476]


<div align="center">

# 5. Линейная алгебра

</div>

---

In [137]:
a = np.array([[2, 1], [2, 3]])
print(a)

[[2 1]
 [2 3]]


In [139]:
# Определитель
a = np.array([[2, 1], [2, 3]])
print(np.linalg.det(a))

4.0


In [140]:
# Обратная матрица
a = np.array([[2, 1], [2, 3]])
b = np.linalg.inv(a)
print(b)

[[ 0.75 -0.25]
 [-0.5   0.5 ]]


In [142]:
# Проверка обратной матрицы, должны быть еденичные матрицы
a = np.array([[2, 1], [2, 3]])
b = np.linalg.inv(a)
print(a.dot(b))
print(b.dot(a))

[[1. 0.]
 [0. 1.]]
[[1. 0.]
 [0. 1.]]


**Решение НЛУ**

$$A * x = v$$

In [143]:
a = np.array([[2, 1], [2, 3]]) 
v = np.array([5, -10])
print(np.linalg.solve(a, v))

[ 6.25 -7.5 ]


**Найти собственные векторы матрицы A**

$$A * x = \lambda * x$$

In [144]:
a = np.array([[2, 1], [2, 3]]) 
l, u = np.linalg.eig(a)
print(l)
print(u)

[1. 4.]
[[-0.70710678 -0.4472136 ]
 [ 0.70710678 -0.89442719]]


**Собственные значения матрицы $A$ и $A^T$ совпадают**

In [145]:
a = np.array([[2, 1], [2, 3]]) 
l, u = np.linalg.eig(a.T)
print(l)
print(u)

[1. 4.]
[[-0.89442719 -0.70710678]
 [ 0.4472136  -0.70710678]]
